### ISSUE TITLE
java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545

### ISSUE THREAD

**Issue by MartinLondonUK77** [veraPDF/veraPDF-library#1197]
This may be a question or a bug as am new to PDF certificates but am moderately familiar with Java.

As java.lang.IllegalAccessError suggests, a method is being called that does not exist.

My problem below is a repeatable issue that occurs in Greenfield 1.18.6 GUI parser.  I have tried running with JDK 17 and 17.01, as well as Eclipse Temurin 17.0.1+12 x64 and gives the same result.

All the files are signed with pdfbox-examples CreateSignature package with adbe.pkcs7.detached - [https://svn.apache.org/viewvc/pdfbox/branches/2.0/examples/src/main/java/org/apache/pdfbox/examples/signature/](url)

The keystore was generated using the same command line keytool as per CreateSignature class: 
keytool -genkeypair -storepass ******** -storetype pkcs12 -alias test -validity 365 -v -keyalg RSA -keystore keystore.p12.

I include demo two files, one produced in Word 2016 and the other raised in PDFBox. Both produce the same error.

PDFBox - [Henery the Eighth PDFBox.PDF](https://github.com/veraPDF/veraPDF-library/files/7460348/Henery.the.Eighth.PDFBox.PDF)
Word - [Henery the Eighth PDFA_signed Word.pdf](https://github.com/veraPDF/veraPDF-library/files/7460357/Henery.the.Eighth.PDFA_signed.Word.pdf)

These demo files are signed as PDF/A but have outstanding issues - I am aware of them.

If someone can look at this and advise I would be grateful. Perhaps I am doing something wrong or it needs a fix.

Here is the dump:-
```
java.util.concurrent.ExecutionException: java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at java.desktop/javax.swing.SwingWorker.get(SwingWorker.java:613)
        at org.verapdf.gui.CheckerPanel.validationEnded(CheckerPanel.java:549)
        at org.verapdf.gui.ValidateWorker.done(ValidateWorker.java:169)
        at java.desktop/javax.swing.SwingWorker$5.run(SwingWorker.java:750)
        at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.run(SwingWorker.java:848)
        at java.desktop/sun.swing.AccumulativeRunnable.run(AccumulativeRunnable.java:112)
        at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.actionPerformed(SwingWorker.java:858)
        at java.desktop/javax.swing.Timer.fireActionPerformed(Timer.java:311)
        at java.desktop/javax.swing.Timer$DoPostEvent.run(Timer.java:243)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:771)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:741)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545
        at org.verapdf.gf.model.impl.external.GFPKCSDataObject.<init>(GFPKCSDataObject.java:59)
        at org.verapdf.gf.model.impl.pd.signature.GFPDSignature.getContents(GFPDSignature.java:86)
        at org.verapdf.gf.model.impl.pd.signature.GFPDSignature.getLinkedObjects(GFPDSignature.java:72)
        at org.verapdf.pdfa.validation.validators.BaseValidator.addAllLinkedObjects(BaseValidator.java:199)
        at org.verapdf.pdfa.validation.validators.BaseValidator.checkNext(BaseValidator.java:166)
        at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:117)
        at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:93)
        at org.verapdf.processor.ProcessorImpl.validate(ProcessorImpl.java:219)
        at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:120)
        at org.verapdf.processor.BatchFileProcessor.processItem(BatchFileProcessor.java:98)
        at org.verapdf.processor.BatchFileProcessor.processList(BatchFileProcessor.java:74)
        at org.verapdf.processor.AbstractBatchProcessor.process(AbstractBatchProcessor.java:102)
        at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:128)
        at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:62)
        at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Nov 02, 2021 11:27:51 AM org.verapdf.gui.utils.DialogUtils errorDialog
SEVERE: Execution exception in processing: java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545
java.util.concurrent.ExecutionException: java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at java.desktop/javax.swing.SwingWorker.get(SwingWorker.java:613)
        at org.verapdf.gui.CheckerPanel.validationEnded(CheckerPanel.java:549)
        at org.verapdf.gui.ValidateWorker.done(ValidateWorker.java:169)
        at java.desktop/javax.swing.SwingWorker$5.run(SwingWorker.java:750)
        at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.run(SwingWorker.java:848)
        at java.desktop/sun.swing.AccumulativeRunnable.run(AccumulativeRunnable.java:112)
        at java.desktop/javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.actionPerformed(SwingWorker.java:858)
        at java.desktop/javax.swing.Timer.fireActionPerformed(Timer.java:311)
        at java.desktop/javax.swing.Timer$DoPostEvent.run(Timer.java:243)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:771)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:741)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.IllegalAccessError: class org.verapdf.gf.model.impl.external.GFPKCSDataObject (in unnamed module @0x17c55545) cannot access class sun.security.pkcs.PKCS7 (in module java.base) because module java.base does not export sun.security.pkcs to unnamed module @0x17c55545
        at org.verapdf.gf.model.impl.external.GFPKCSDataObject.<init>(GFPKCSDataObject.java:59)
        at org.verapdf.gf.model.impl.pd.signature.GFPDSignature.getContents(GFPDSignature.java:86)
        at org.verapdf.gf.model.impl.pd.signature.GFPDSignature.getLinkedObjects(GFPDSignature.java:72)
        at org.verapdf.pdfa.validation.validators.BaseValidator.addAllLinkedObjects(BaseValidator.java:199)
        at org.verapdf.pdfa.validation.validators.BaseValidator.checkNext(BaseValidator.java:166)
        at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:117)
        at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:93)
        at org.verapdf.processor.ProcessorImpl.validate(ProcessorImpl.java:219)
        at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:120)
        at org.verapdf.processor.BatchFileProcessor.processItem(BatchFileProcessor.java:98)
        at org.verapdf.processor.BatchFileProcessor.processList(BatchFileProcessor.java:74)
        at org.verapdf.processor.AbstractBatchProcessor.process(AbstractBatchProcessor.java:102)
        at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:128)
        at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:62)
        at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by bdoubrov** [comment]
This error is caused by the fact that `sun.security` module we use to parse PKCS7-based digital signatures is no longer supported in Java 16 and higher. We'll add Java 16 support in the coming veraPDF release by using the methods described here: https://openjdk.java.net/jeps/396 . But unfortunately Java 17 support will be postponed till later. 

You may still use veraPDF 1.18 with any JDK version below 16.

**Comment by MartinLondonUK77** [comment]
Thanks for explaining, I can confirm that JDK 15 works well with veraPDF 1.18.

I look forward to future releases in due course.

**Comment by bdoubrov** [comment]
Thanks for the confirmation. 

The latest dev build https://software.verapdf.org/develop/1.19/verapdf-greenfield-1.19.229-installer.zip seems to fix this issue for JDK 17 as well by explicitly exporting `sun.security.pkcs` module: https://github.com/veraPDF/veraPDF-apps/commit/5ad644aeb8a18dfefd78f5084d8037b4dfb2f307  . Feel free to check if this indeed works for you.

**Comment by MartinLondonUK77** [comment]
Yes, JDK 17 does appear to resolve the issue. Thanks for your help.

**Comment by bsanchezb** [comment]
Hello @bdoubrov ,

After upgrading to VeraPdf 1.20.1, it seems like the issue is still present when validating PDF containing a PKCS#7 signature.
The issue does not appear when using JDK 15 or lower, but I obtain the same exception, as in the topic's description, on JDK 16 and 17.

Could you please take a look on the file within the attachment?

Best regards,
Aleksandr.

[testdoc-signed.pdf](https://github.com/veraPDF/veraPDF-library/files/8187226/testdoc-signed.pdf)

**Comment by bdoubrov** [comment]
We have added the greenfield PKCS7 parser that replaces the sun.security.pkcs implementation. Feel free to test it with the latest dev build available at: https://software.verapdf.org/dev/verapdf-installer.zip

**Comment by bsanchezb** [comment]
Confirm the issue is fixed in 1.22.2.
Tested even with JDK 18, works as expected!
